Added campaign_name field to Donation model

olasitarska opened the pull request.

jezdez commented on fundraising/views.py:
```diff
@@ -77,10 +82,11 @@ def donate(request):
         if fixed_amount:
             try:
-                initial = {'amount': Decimal(fixed_amount)}
+                initial = {'amount': Decimal(fixed_amount), 'campaign': campaign}
```
Would you mind reformatting that to multiple lines?

timgraham commented on fundraising/views.py:
```diff
@@ -27,16 +27,19 @@ def index(request):
         'testimonial': Testimonial.objects.filter(is_active=True).order_by('?').first(),
         'display_logo_amount': DISPLAY_LOGO_AMOUNT,
         'weekly_goal': WEEKLY_GOAL,
+        'campaign': campaign,
```
is this used in the template?

jezdez commented on fundraising/views.py:
```diff
@@ -27,16 +27,19 @@ def index(request):
         'other_donors': other_donors,
         'total_donors': DjangoHero.objects.count(),
-        'form': DonateForm(initial={'amount': DEFAULT_DONATION_AMOUNT}),
+        'form': DonateForm(initial={'amount': DEFAULT_DONATION_AMOUNT, 'campaign': campaign}),
```
Also multiple lines please

timgraham commented on fundraising/views.py:
```diff
@@ -70,6 +74,7 @@ def donate(request):
                     amount=amount,
                     stripe_charge_id=charge.id,
                     stripe_customer_id=customer.id,
+                    campaign_name=campaign
```
please include a trailing comma so if more items are added, we don't have to modify the line again

timgraham commented on fundraising/views.py:
```diff
@@ -77,10 +82,11 @@ def donate(request):
     else:
         fixed_amount = request.GET.get('amount') or None
-        initial = {}
+        campaign = request.GET.get('campaign') or None
```
`or None` isn't needed as `.get('nonexistent'`) returns `None`

jezdez replied:
Unless it's an empty string

jezdez replied:
Like /foundraising/donate/?campaign=

timgraham replied:
I like using `None` here might result in `campaign_name='None'` though.

timgraham commented on fundraising/views.py:
```diff
@@ -27,16 +27,19 @@ def index(request):
     donors_with_logo = DjangoHero.objects.in_period(begin, end, with_logo=True)
     other_donors = DjangoHero.objects.in_period(begin, end)
 
+    campaign = request.GET.get('campaign', None)
```
`, None` not needed

timgraham commented on fundraising/forms.py:
```diff
@@ -160,6 +161,11 @@ class PaymentForm(forms.Form):
 
+    campaign = forms.CharField(
+        required=False,
+        widget=forms.HiddenInput()
```
this could fit on 1 line as above, otherwise please include a trailing comma as noted below

jezdez commented:
LGTM

Of course if we want to store more info about campaigns than just the name maybe it'd be better to have a separate campaign data model and have an optional foreign key in the donation model. That allows us to have in theory easy metadata storage and is future proof. E.g. imagine having a campaign page with just the heart for that campaign. In fact you could consider this fundraiser the first campaign we could use this system for.

timgraham commented on fundraising/forms.py:
```diff
@@ -160,6 +161,11 @@ class PaymentForm(forms.Form):
         ),
     )
 
```
no newline between fields (you can remove the one above `expires` please)

timgraham commented:
The complexity of code without tests is starting to make me nervous. Would you like to write some? :tests: <-- where is a good icon when I need it?

jezdez commented:
There is is: :goat:

jezdez commented:
Reference (et al): http://www.obeythetestinggoat.com/

jezdez commented on fundraising/views.py:
```diff
@@ -70,17 +76,22 @@ def donate(request):
             if 'amount' in form.errors:
                 show_amount = True
     else:
-        fixed_amount = request.GET.get('amount') or None
```
Please don't do that, it will break with empty GET parameters.

jezdez commented on fundraising/views.py:
```diff
@@ -70,17 +76,22 @@ def donate(request):
-        fixed_amount = request.GET.get('amount') or None
-        initial = {}
+        fixed_amount = request.GET.get('amount')
+        campaign = request.GET.get('campaign')
```
As I said before, this needs to be request.GET.get('campaign') or None since that also catches empty strings

timgraham replied:
Is an empty string a problem for `campaign` though? That's the "empty" value for a blank character field.

timgraham commented on fundraising/models.py:
```diff
@@ -75,6 +75,7 @@ class Donation(FundraisingModel):
     donor = models.ForeignKey(DjangoHero, null=True)
     stripe_charge_id = models.CharField(max_length=100, null=True)
     stripe_customer_id = models.CharField(max_length=100, null=True)
+    campaign_name = models.CharField(max_length=100, null=True)
```
It's better not to use `null` for `CharField` because then there's two "empty values" "NULL" and "empty string" -- this should have `blank=True` though.

olasitarska commented:
I pushed some tests here.. the submitting donation form test doesn't work yet, mocking is not really working as it should (maybe I used wrong url?). I am too tired to work on this longer today though. Not sure about tomorrow yet. Sorry!

jezdez commented on fundraising/tests.py:
```diff
@@ -1,10 +1,73 @@
+
+        response = self.client.post(reverse('fundraising:donate'), {
+            'amount': 100,
+            'number': '4242424242424242',
```
The number, cvc and expires data is not submitted to our form view, the form actually only submits the `stripe_token` that is set by our JavaScript. Card data never hits our servers :) So simply set a `stripe_token` to something and it should be valid.

jezdez commented on fundraising/tests.py:
```diff
@@ -1,10 +1,73 @@
+    @requests_mock.mock()
+    def setUpClass(cls, mocker):
+        mocker.register_uri(
+            'GET',
```
I'm pretty sure this should be POST, right?

jezdez commented on fundraising/tests.py:
```diff
@@ -1,10 +1,73 @@
+
     def test_donors_count(self):
         DjangoHero.objects.create()
         response = self.client.get('/fundraising/')
```
Use `reverse('fundraising:index')` here?

jezdez commented on fundraising/tests.py:
```diff
@@ -1,10 +1,73 @@
+        self.assertEqual(response.context['publishable_key'], settings.STRIPE_PUBLISHABLE_KEY)
+
+        # Checking if amount field is hidden
+        self.assertEqual(response.context['form'].fields['amount'].widget.__class__, forms.HiddenInput)
```
use `self.assertIsInstance` please

jezdez commented on fundraising/tests.py:
```diff
@@ -1,10 +1,73 @@
+        self.assertEqual(response.context['publishable_key'], settings.STRIPE_PUBLISHABLE_KEY)
+
+        # Checking if amount field is visible
+        self.assertEqual(response.context['form'].fields['amount'].widget.__class__, forms.TextInput)
```
also `self.assertIsInstance`

bmispelon replied:
what about `assertFalse(response.context['form'].fields['amount'].widget.is_hidden)`?

jezdez commented on fundraising/tests.py:
```diff
@@ -1,10 +1,73 @@
+
+    def test_render_donate_form_with_campaign(self):
+        campaign = 'sample'
+        response = self.client.get(reverse('fundraising:donate'), {'amount': 100, 'campaign':campaign})
```
space after colon

jezdez commented on fundraising/tests.py:
```diff
@@ -1,10 +1,73 @@
+        response = self.client.get(reverse('fundraising:donate'), {'amount': 100, 'campaign':campaign})
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.context['fixed_amount'], '100')
+        self.assertEqual(response.context['publishable_key'], settings.STRIPE_PUBLISHABLE_KEY)
```
this misses a test for the case in which the `campaign` GET variable is an empty string, like when doing `/fundraising/donate/?campaign=`

jezdez commented on fundraising/tests.py:
```diff
@@ -1,10 +1,73 @@
+            'campaign': None,
+            'stripe_token': 'test',
+        })
+        self.assertTrue(response.context['form'].is_valid())
```
This test should also test for the existence of an Donate object on successful POST

jezdez commented on fundraising/tests.py:
```diff
@@ -1,10 +1,73 @@
+            'GET',
+            'https://api.stripe.com/v1/customers',
+            status_code=200
+        )
```
This may need more mocking, I think we use more than just the customer API endpoints, but also charge. See the logs in the stripe dashboard for a clear idea which endpoints are used

olasitarska commented:
I fixed all the things you commented on, thanks!

I am still not able to mock Stripe requests despite mocking all of the urls. Looking [at this](http://vincent.is/testing-django-projects/), there is definitely more magic involved than just trying to mock the request with `request_mock`, I think.

olasitarska commented:
Uh finally! For some reason mocking API calls via setUpClass didn't work *at all*. It only worked when I added the mocking decorator directly on the test. Anyway, then together with @bmispelon we've decided that mocking API calls isn't really a good idea, because Stripe wanted us to return a very particular response. So these last tests I added are patching Stripe methods and it works. 

Is this all ok?

asendecka commented:
Meanwhile I and @oinopion took a look and refactor `donate` method a little bit. Previous state was pretty hard to test. We moved stripe & creating of a donation to form method. We also added logging, since there was only `raise` there. For now we simply use Bad Request. I think it should be addressed when solving #342 (maybe displaying form again?), but it's better than 500 anyway.

jezdez commented:
@asendecka The reason why I fell back to just catching `StripeError` was simply time constraints on the implementation (I did that in one afternoon after dabbling with the various Django apps that didn't do what we want), but I know there are more than just this one exception. A bad request is only better for the user, but letting the exception bubble up allowed us to see what happened in Sentry, which we wouldn't with the 400. Logging doesn't help as much here since we don't have a sensible logging setup.

Check out `CardError` (https://github.com/stripe/stripe-python/blob/846ec9080a7f3cd55113871803eb9d7c29e869d7/stripe/error.py#L29-L36) that has a status code attached with the reason why the card was declined: https://stripe.com/docs/api/python#errors

There is also a pretty awesome example for how to catch all the other exceptions in Python below that list errors in the Stripe docs, under "Handling errors".

  try:
    # Use Stripe's bindings...
    pass
  except stripe.error.CardError, e:
    # Since it's a decline, stripe.error.CardError will be caught
    body = e.json_body
    err  = body['error']

    print "Status is: %s" % e.http_status
    print "Type is: %s" % err['type']
    print "Code is: %s" % err['code']
    # param is '' in this case
    print "Param is: %s" % err['param']
    print "Message is: %s" % err['message']
  except stripe.error.InvalidRequestError, e:
    # Invalid parameters were supplied to Stripe's API
    pass
  except stripe.error.AuthenticationError, e:
    # Authentication with Stripe's API failed
    # (maybe you changed API keys recently)
    pass
  except stripe.error.APIConnectionError, e:
    # Network communication with Stripe failed
    pass
  except stripe.error.StripeError, e:
    # Display a very generic error to the user, and maybe send
    # yourself an email
    pass
  except Exception, e:
    # Something else happened, completely unrelated to Stripe
    pass

I think in case a `CardError` is raised we should reraise it as a ValidationError with saying something like `"We're sorry but we had problems charging your card: %s" % errror_code`. `InvalidRequestError` and `AuthenticationError` should raise a `ImproperlyConfigured` exception (with a 500 so it shows up in Sentry) since that's an error on our side, `APIConnectionError` probably should just redisplay the form and ask the user to try again. And then the raw `StripeError` and `Exception` should just be raised to bubble up in Sentry.

Now that you've already started the refactor, let's merge #342 into this and refactor it completely.

olasitarska commented:
Are going to be able to merge this small feature this pull request was about before this campaign ends?

jezdez commented:
Merged in 9e9d176cd6f53a32f0e957a5631dc3b898211370.
